Add basic p-refinement criterion based on truncation error

nilsvu opened the pull request:
## Proposed changes

Just use the `truncation_error` estimate based on power monitors to p-refine to a target error. We can improve and expand this as we go. This p-refines the `ExportCoordinates2D` executable already pretty well when using the inertial coordinates to compute truncation errors and a Sphere domain in the input file.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Add basic p-refinement criterion based on truncation error

nilsvu committed:
Use truncation error AMR criterion in ExportCoordinates

Monitor truncation error of the coordinates.

nilsdeppe commented on src/ParallelAlgorithms/Amr/Criteria/TruncationError.hpp:
```diff
@@ -0,0 +1,113 @@
+
+  TruncationError() = default;
+
+  TruncationError(const double target_truncation_error)
```
Move out of line

nilsdeppe commented on src/ParallelAlgorithms/Amr/Criteria/TruncationError.hpp:
```diff
@@ -0,0 +1,113 @@
+  using argument_tags = tmpl::list<::Tags::DataBox>;
+
+  template <typename DbTagsList, typename Metavariables>
+  auto operator()(const db::DataBox<DbTagsList>& box,
+                  Parallel::GlobalCache<Metavariables>& /*cache*/,
+                  const ElementId<Dim>& /*element_id*/) const {
+    auto result = make_array<Dim>(Flag::DoNothing);
```
Move out of line

Don't use auto return type if you know the return type

nilsdeppe commented on src/ParallelAlgorithms/Amr/Criteria/TruncationError.hpp:
```diff
@@ -0,0 +1,113 @@
+    tmpl::for_each<TensorTags>(
+        [&result, &box, &mesh,
+         target_truncation_error = target_truncation_error_](const auto tag_v) {
+          // Continue only if AMR flags are still unset
```
Why? should there not be some sort of priority procedure over all the flags? If not, this should be documented.

nilsvu replied:
Added with the coarsening

nilsdeppe commented on src/ParallelAlgorithms/Amr/Criteria/TruncationError.hpp:
```diff
@@ -0,0 +1,113 @@
+template <size_t Dim, typename TensorTags>
+class TruncationError : public Criterion {
+ public:
+  struct TargetTruncationError {
```
Name this `TargetAbsoluteTruncationError`?

Why not include a relative error too?

nilsvu replied:
Added relative and absolute error

kidder commented on src/ParallelAlgorithms/Amr/Criteria/TruncationError.hpp:
```diff
@@ -0,0 +1,113 @@
+  auto operator()(const db::DataBox<DbTagsList>& box,
+                  Parallel::GlobalCache<Metavariables>& /*cache*/,
+                  const ElementId<Dim>& /*element_id*/) const {
+    auto result = make_array<Dim>(Flag::DoNothing);
```
So the way this criterion is written, AMR will never be able to decrease the resolution as returning DoNothing will always take priority over any criteria that wants to DecreaseResolution.  So this criterion should handle when to request to decrease the resolution as well...

nilsvu replied:
Can I return `Undefined` to signal that this criterion makes no statement about what should happen in that dimension?

nilsdeppe replied:
How can this ever not have an opinion on what to do? The error is either above or below tolerances.

nilsvu replied:
Currently it doesn't try to coarsen the resolution, but I can add that. There should probably be a safety factor though or an acceptable absolute error range so it doesn't switch back and forth between resolutions at every AMR step?

nilsvu replied:
As a first pass I was going to ignore coarsening. We have to experiment with this anyway.

nilsdeppe replied:
Let's just have a meet about this. That's easier than going back and forth on gitHub comments :)

kidder replied:
As AMR is currently implemented, you should return `Join` (the lowest priority flag) if you do not want to override the decisions of other criteria.  But as this is a p-refinement criterion, it should return one of `{IncreaseResolution, DoNothing, DecreaseResolution}` I would have expected this criterion to compute a third number besides the error and target error, such as the estimated error dropping the top mode, in order to determine if the resolution can be decreased.

kidder commented on src/ParallelAlgorithms/Amr/Criteria/TruncationError.hpp:
```diff
@@ -107,7 +107,9 @@ class TruncationError : public Criterion {
   double target_truncation_error_{};
 };
 
+/// \cond
```
the documentation wrappers should be in the previous commit

nilsvu commented:
Rewrote the implementation to work with relative and absolute errors, and do refinement+coarsening.

nilsdeppe commented on src/ParallelAlgorithms/Amr/Criteria/TruncationError.hpp:
```diff
@@ -0,0 +1,212 @@
+ * \tparam Dim Spatial dimension of the grid
+ * \tparam TensorTags List of tags of the tensors to be monitored
+ */
+template <size_t Dim, typename TensorTags>
```
[optional] Maybe in the future we will want to make it so that we can runtime choose how to deal with the tensor tags? E.g. we might want to monitor the L2 norm over the spacetime metric components, etc. It would also be nice to be able to have some runtime control over which of the tags we monitor, so we don't need to recompile to monitor different tags. E.g. we could just add all evolved and primitive tags to the list and then decide at runtime which we actually deal with. Again, not for this PR, but wanted to mention it and get your thoughts. I could see this being especially useful during testing/debugging times of the criterion.

nilsvu replied:
Agreed. Added the runtime control over which variables to monitor. I left out the norms over tensor components for now (not sure if we really need that because monitoring the L2 norm over something like the metric seems pretty much the same as monitoring each component. We can add this later if we want it.)

nilsdeppe replied:
Sounds good. The advantage of monitoring an L2 norm over components would be fewer matrix multiplies :) But we can experiment with this in the futer.

nilsdeppe commented on src/ParallelAlgorithms/Amr/Criteria/TruncationError.hpp:
```diff
@@ -0,0 +1,212 @@
+    }
+    using tag = tmpl::type_from<std::decay_t<decltype(tag_v)>>;
+    const auto& tensor = db::get<tag>(box);
+    for (const DataVector& tensor_component : tensor) {
```
what do you think about moving the work in a `cpp` file since this just acts on one DataVector? We just do _so_ much in hpp files and then tweaking anything means recompiling the executable. It would be nice for testing to be able to tweak the algorithm in the cpp file to only have to link. I think we could even extend to taking `Tensor::const_iterator` types and then do the loop over components in the cpp. This would mean that even things like L2 norms of components could be added and debugged with mostly only changing cpp files.

nilsvu replied:
Moved to the cpp. Left the loop over components for now (not trivial because `Tensor::const_iterator` is templated and not sure if we need norms over components at all. Can be done later if needed.)

nilsdeppe commented on src/ParallelAlgorithms/Amr/Criteria/TruncationError.hpp:
```diff
@@ -0,0 +1,212 @@
+              umax * rel_truncation_error_coarsened;
+          if ((target_rel_truncation_error.has_value() and
+               rel_truncation_error_coarsened <=
+                   target_rel_truncation_error.value()) and
```
can you use `.value_or(0.0)` or `.value_or(negative_infinity)`? 2x

nilsvu replied:
Tried it but I think it makes the logic a bit harder to understand because the two conditions for refine/coarsen have to use different `value_or` values.

nilsdeppe reviewed (changes requested):
LGTM, a couple requests. You can squash immediately.

kidder reviewed (commented):
I have no requested changes

nilsvu committed:
Add validate_selection function

Used in a couple of places to check that the given names
in an input file option are valid tags.

nilsvu commented:
Pushed new first commit and a fixup

nilsdeppe reviewed (commented):
LGTM, please squash
